### DISCUSSION TITLE
Well Formed XML Log Output

### DISCUSSION THREAD

**Discussion by CharityKathure** [microsoft/windows-container-tools#123] [Ideas]
Going forward, users will be able to choose between displaying windows container logs in JSON or XML. To view the implementation, refer to this branch: https://github.com/microsoft/windows-container-tools/tree/v2/main-xml

The legacy XML output is not well formed. i.e. (Event Log): 
```
<Source>EventLog</Source>
<LogEntry>
    <Time>...</Time>
    <Channel>...</Channel>
    <Level>...</Level>
    <EventId>...</EventId>
    <Message>...</Message>
</LogEntry>
```

**For the XML output to be well formed, it needs to have a root element.**

**Proposed XML format:**
Introducing a new root element _`<log>`_ to the current format.
```
<log>
    ...
    ...
</log>
```

Event Log:
```
<Log>
    <Source>EventLog</Source>
    <LogEntry>
        <Time>....</Time>
        <Channel>...</Channel>
        <Level>...</Level>
        <EventId>...</EventId>
        <Message>...</Message>
    </LogEntry>
</Log>
```

ETW Log:
```
<Log>
    <Source>ETW</Source>
    <LogEntry>
        <Time>...</Time>
        <ProviderName>...</ProviderName>
	<ProviderId>...</ProviderId>
	<DecodingSource>...</DecodingSource>
	<Execution>
		<ProcessId>...</ProcessId>
		<ThreadId>...</ThreadId>
	</Execution>
	<Level>...</Level>
	<Keyword>...</Keyword>
	<EventId>...</EventId>
	<EventData>
		<FrameUniqueID>...</FrameUniqueID>
		<PortNumber>...</PortNumber>
		<TID>...</TID>
		<PeerID>...</PeerID>
		<PayloadLength>...</PayloadLength>
		<QueueLength>...</QueueLength>
		<QueueState>...</QueueState>
	</EventData>
    </LogEntry>
</Log>
```

Process Log:
```
<Log>
    <Source>Process</Source>
    <LogEntry>
        <Logline>...</Logline>
    </LogEntry>
</Log>
```

File Log:
```
<Log>
    <Source>File</Source>
    <LogEntry>
        <Logline>...</Logline>
	<FileName>*.log</FileName>
    </LogEntry>
</Log>
```

**Comment by CharityKathure** [comment]
cc @profnandaa, @iankingori, @bobsira

**Reply by profnandaa** [reply]
The schema proposal looks good to me, I don't have any additions to make.

**Comment by CharityKathure** [comment]
cc @nelljerram, @lucillex

**Comment by bobsira** [comment]
Did we have any discussions on the preamble tag that included the xml version, encoding and namespace?

**Comment by CharityKathure** [comment]
Closing this discussion. The feature is available in [Log Monitor Version 2.1.0](https://github.com/microsoft/windows-container-tools/releases/tag/v2.1.0)
